File: demo/RedBlack/redblack.f
      PROGRAM REDBF
      PARAMETER (N=10)
      REAL A(N,N), EPS, MAXEXP, W
      INTEGER ITMAX
CDVM$ DISTRIBUTE A(BLOCK, BLOCK)
      PRINT *,  '**********  TEST_REDBLACK   **********'
      ITMAX = 20
      MAXEXP = 0.5E - 5
      W = 0.5
CDVM$ PARALLEL (J,I) ON A(I, J)
      DO 1  J = 1,N
      DO 1  I = 1,N
         IF (I.EQ.J) THEN
           A(I,J) = N+2
         ELSE
          A(I,J) = -1. 
         ENDIF
1     CONTINUE
      DO 2  IT = 1, ITMAX
      EPS = 0.
C	loop for red and black variables 
      DO 3 IRB = 0,1
CDVM$ PARALLEL (J,I) ON A(I, J), NEW (S), REDUCTION (MAX(EPS)),
CDVM$*                      SHADOW_RENEW  (A)
C	variable S - private variable in loop iterations
C	variable EPS is used for calculation of maximum value 

C	Exception : iteration space is not rectangular

      DO 21  J = 2,N-1
      DO 21  I = 2 + MOD(J+IRB,2), N-1, 2
         S = A(I,J)
         A(I,J) = (W/4) * (A(I-1,J) + A(I+1,J) + A(I,J-1) +
     *   A(I,J+1)) + (1-W) * A(I,J)
         EPS = MAX (EPS, ABS(S - A(I,J)))
21    CONTINUE
3     CONTINUE
      PRINT 200, IT, EPS
200   FORMAT(' IT = ',I4, '   EPS = ', E14.7)
      IF (EPS.LT.MAXEXP) GO TO 4
2     CONTINUE
4     OPEN (3, FILE='REDBF.DAT', FORM='FORMATTED',STATUS='UNKNOWN')
      WRITE (3,*)   A
      CLOSE (3)
      END

File: examples/RedBlack/redblack.c
/* RED-BLACK program */

#include <math.h>
#include <stdlib.h>
#include <stdio.h>

#define Max(a, b) ((a) > (b) ? (a) : (b))

#define N 8
#define ITMAX 10

int main(int an, char ** as)
{
    int i, j, it;
    float MAXEPS = 0.5E-5f;
    float w = 0.5f;

    #pragma dvm array distribute[block][block]
    float (*A)[N];

    /* Create array */
    A = malloc(N * N * sizeof(float));

    #pragma dvm region
    {
    /* Initialization parallel loop */
    #pragma dvm parallel([i][j] on A[i][j]) cuda_block(256)
    for (i = 0; i < N; i++)
        for (j = 0; j < N; j++)
            if (i == j)
                A[i][j] = N + 2;
            else
                A[i][j] = -1.0f;
    }

    /* iteration loop */
    for (it = 1; it <= ITMAX; it++)
    {
        float eps = 0.f;
        #pragma dvm actual(eps)

        #pragma dvm region
        {
        /* Parallel loop with reduction on RED points */
        #pragma dvm parallel([i][j] on A[i][j]) shadow_renew(A), reduction(max(eps)), cuda_block(256)
        for (i = 1; i < N - 1; i++)
            for (j = 1; j < N - 1; j++)
                if ((i + j) % 2 == 1)
                {
                    float s;
                    s = A[i][j];
                    A[i][j] = (w / 4) * (A[i - 1][j] + A[i + 1][j] + A[i][j - 1] + A[i][j + 1]) + (1 - w) * A[i][j];
                    eps = Max(fabs(s - A[i][j]), eps);
                }

        /* Parallel loop on BLACK points (without reduction) */
        #pragma dvm parallel([i][j] on A[i][j]) shadow_renew(A), cuda_block(256)
        for (i = 1; i < N - 1; i++)
            for (j = 1; j < N - 1; j++)
                if ((i + j) % 2 == 0)
                {
                    A[i][j] = (w / 4) * (A[i - 1][j] + A[i + 1][j] + A[i][j - 1] + A[i][j + 1]) + (1 - w) * A[i][j];
                }
        }

        #pragma dvm get_actual(eps)
        printf("it=%4i   eps=%e\n", it, eps);
        if (eps < MAXEPS)
            break;
    }

    free(A);
    return 0;
}
